### PR TITLE
fix(fmt): Preserve multiple newlines between elements (#374)

### PR DESCRIPTION
**Pull request by AlexanderArvidsson**
Preserves multiple lines between elements, to allow organizing code better.
For reference and discussion, see #374.

`templ fmt` [is inconsistent with itself](https://github.com/a-h/templ/issues/374#issuecomment-2027357343), sometimes preserving newlines (collapsing) and sometimes now.
This aligns `templ fmt` with `gofmt`, which also preserves newlines (by collapsing multiple standalone newlines into one), which `templ fmt` also does in go functions.

### CODE REVIEW

File: parser/v2/types.go [hunk 1]
```diff
@@ -375,17 +375,27 @@ func (t HTMLTemplate) Write(w io.Writer, indent int) error {
...
+	SpaceNone           TrailingSpace = ""
+	SpaceHorizontal     TrailingSpace = " "
+	SpaceVertical       TrailingSpace = "\n"
+	SpaceVerticalDouble TrailingSpace = "\n\n"
 )
 
 var ErrNonSpaceCharacter = errors.New("non space character found")
 
-func NewTrailingSpace(s string) (ts TrailingSpace, err error) {
+func NewTrailingSpace(s string, allowMulti bool) (ts TrailingSpace, err error) {
```
Comment (joerdav) [inline parser/v2/types.go:386]: This function is getting quite "indented", is there any refactors we can make here to make the flow easier to follow?

Comment (joerdav) [reply parser/v2/types.go:386]: Possibly could be a different function `NewTrailingSpace` and `NewMultilineTrailingSpace`. Rather than a flag. Would that potentially make it cleaner, or just add duplication?

**Review by joerdav — CHANGES_REQUESTED** [review]

File: parser/v2/types.go (hunk 1 above)
Comment (AlexanderArvidsson) [reply parser/v2/types.go:386]: The majority of that function would be the same between the two, so you'd end up either duplicating the code, or finding a different type of abstraction. I'd say this solution is the least "abstraction-hell" and at this point I wouldn't worry too much with it.
But I agree if more parameters are introduced, its responsibility would be too large, but not really in its current state I think.

**Comment by AlexanderArvidsson** [comment]
I found a case that I'd like to include before potentially accepting this PR:
```
			</header>

			@c.Hero()
			<main class="max-w-7xl mx-auto px-8 mt-24 overflow-hidden">
```
The space should have been preserved between `c.Hero` and `main`, but it was stripped.

**Comment by AlexanderArvidsson** [comment]
I also found that this apparently throws an error, due to the first `    <a></a>`. Says `test_templ.go:32:61: string literal not terminated (and 8 more errors)`.
I will check it out.
```
package components

templ Test() {
  <div class="flex items-center gap-4">
    <a></a>



    <a>
    </a>
  </div>
}
```

**Comment by AlexanderArvidsson** [comment]
@joerdav Before I go and preserve newlines for statements like `if` and `for`, I've decided to only focus on Elements and TemplElements.
Adding the trailing space logic to every statement is a bigger change that I think would only be fair if I got your opinion on first.

You can see how I added the logic to `TemplElementExpression` and let me know if I should do the same for `if`, `for`, etc.

I believe consistency is important, so I think preserving newlines should work as you expect for all types of statements or expressions, but let me know your thoughts.

**Comment by joerdav** [comment]
Hey @AlexanderArvidsson , thanks for all the time on this so far, I've looked at the code here and am happy with how you have approached this. I agree with you that due to consistency I think that we should support trailing spaces for all elements.

I think going ahead with this pattern is a solid approach.

One to think about is that the following code block may end up repeated a bit so could maybe be made into a function (or not since I suppose it isn't a lot of code):

```

		// Parse trailing whitespace after expression.
		ws, _, err := parse.Whitespace.Parse(pi)
		if err != nil {
			return r, false, err
		}
		r.TrailingSpace, err = NewTrailingSpace(ws, true)
		if err != nil {
			return r, false, err
		}
```

**Comment by AlexanderArvidsson** [comment]
> Hey @AlexanderArvidsson , thanks for all the time on this so far, I've looked at the code here and am happy with how you have approached this. I agree with you that due to consistency I think that we should support trailing spaces for all elements.
> 
> I think going ahead with this pattern is a solid approach.
> 
> One to think about is that the following code block may end up repeated a bit so could maybe be made into a function (or not since I suppose it isn't a lot of code):
> 
> ```
> 
> 		// Parse trailing whitespace after expression.
> 		ws, _, err := parse.Whitespace.Parse(pi)
> 		if err != nil {
> 			return r, false, err
> 		}
> 		r.TrailingSpace, err = NewTrailingSpace(ws, true)
> 		if err != nil {
> 			return r, false, err
> 		}
> ```

The `elementparser` has a function `addTrailingSpaceAndValidate` which does _almost_ the same thing, it also checks for `voidElementCloser` and then validates. I can put the `addTrailingSpace` portion of it in its own function and use that everywhere!

So far, I just followed the "don't repeat yourself more than twice" philosophy, and just copied that small piece of code twice. But for more uses I would definitely put it in a function.

**Comment by AlexanderArvidsson** [comment]
@joerdav I've went ahead and added trailing space logic to most parsers and added relevant tests.
It's quite a lot of additions, but they're all very similar.

Please take an extra look at how I unified the trailing space logic in `addTrailingSpace` via the interface `TrailingSpaceSetter`. I kept them separate to the WhitespaceTrailer due to the required pointer receiver, but there might be a better way.

Other than that, from my point of view, this is ready to be tested!

**Comment by joerdav** [comment]
Amazing, thanks! I've tested this out on a fairly large repo at my company and as you said there is no effect to existing code.

I also have read the trailing space interface logic, I think it is done sensibly!

One note I have is on the `addTrailingSpace` function. I can't see any usages of 2 of the return values. Which could also mean that this no longer needs to be a generic function: `func addTrailingSpace(e TrailingSpaceSetter, pi *parse.Input, allowMulti bool) error`.

Going further, I wonder if we need the interface at all? If the trailing space function looked like: `func parseTrailingSpace(pi *parse.Input, allowMulti bool) (types.TrailingSpace, err error)`

It could be used as:

```
	// Parse trailing whitespace after closing brace.
	r.TrailingSpace, err = parseTrailingSpace(pi, true)
	if err != nil {
		return r, false, err
	}
```

What do you think?

Also in general, I think it would be worth @a-h casting his eyes over this.

**Comment by AlexanderArvidsson** [comment]
@joerdav @a-h This should be ready for a final look-through now! I've applied the previous suggestions, even went a bit further for the statements that enforce a breaking space afterwards (like if, for, switch, etc).

**Comment by AlexanderArvidsson** [comment]
@joerdav I've fixed the tests, although I would highly suggest checking the last commit, because I'm not sure if it's correct.
If, switch and for are treated as inline, which to me is weird. To fix the tests introducing double whitespace, I had to only write whitespace when the current node is not if, switch or for, but keep whitespace logic if the _next_ node is if, switch or for. Weird one, but it fixed the tests.

File: generator/test-import/template_templ.go [hunk 2]
```diff
@@ -136,10 +136,6 @@ func main() templ.Component {
 			if templ_7745c5c3_Err != nil {
 				return templ_7745c5c3_Err
 			}
-			_, templ_7745c5c3_Err = templ_7745c5c3_Buffer.WriteString(" ")
-			if templ_7745c5c3_Err != nil {
-				return templ_7745c5c3_Err
-			}
 			templ_7745c5c3_Var6 := templruntime.GeneratedTemplate(func(templ_7745c5c3_Input templruntime.GeneratedComponentInput) (templ_7745c5c3_Err error) {
 				templ_7745c5c3_W, ctx := templ_7745c5c3_Input.Writer, templ_7745c5c3_Input.Context
 				templ_7745c5c3_Buffer, templ_7745c5c3_IsBuffer := templruntime.GetBuffer(templ_7745c5c3_W)
```
Comment (AlexanderArvidsson) [inline generator/test-import/template_templ.go:141]: About these changes, I'd argue this is _more_ correct, since the default rendering does not include these whitespace.

Consider the following:
```templ
templ DashboardRoute(str string) {
	<button>End</button>

	<ul>
		<li>
			<u>Item 1</u>
		</li>
		<li>
			<u>Item 2</u>
		</li>
		<li>
			<u>Item 3</u>
		</li>
	</ul>

	<br/>

	@list() {
		@listItem() {
			<u>Item 1</u>
		}
		@listItem() {
			<u>Item 2</u>
		}
		@listItem() {
			<u>Item 3</u>
		}
	}
}
```

Before, this was generated to:
```html
<button>End</button><ul><li><u>Item 1</u></li><li><u>Item 2</u></li><li><u>Item 3</u></li></ul><br><ul><li><u>Item 1</u></li> <li><u>Item 2</u></li> <li><u>Item 3</u></li></ul>
```
Notice the spacing on the last ones, after the `<br>`

With these changes:
```html
<button>End</button><ul><li><u>Item 1</u></li><li><u>Item 2</u></li><li><u>Item 3</u></li></ul><br><ul><li><u>Item 1</u></li><li><u>Item 2</u></li><li><u>Item 3</u></li></ul>
```

There shouldn't be any spaces between the `<li>` elements, there doesn't need to be here.
Let me know your thoughts on this. I'm not sure if this affects existing code-bases!

**Review by a-h — CHANGES_REQUESTED** [review]
Impressive changes. Thanks a lot for all the effort you've put in to this.

It sounds easy to get the formatting spacing right, but somehow, it isn't at all! The tests, in particular, look great.

The changes I've asked for are really just to make the spacing algorithm clearer to understand for anyone that needs to work in this area in the future.

If you're not able to make the changes due to time constraints, or you're just fed up, let me know and I can refactor from your starting position and merge your PR.

Thanks again!

File: generator/generator.go [hunk 3]
```diff
@@ -540,14 +543,25 @@ func (g *generator) writeNode(indentLevel int, current parser.Node, next parser.
...
 		err = g.writeRawElement(indentLevel, n)
 	case parser.ForExpression:
+		maybeWhitespace = false
 		err = g.writeForExpression(indentLevel, n, next)
 	case parser.CallTemplateExpression:
 		err = g.writeCallTemplateExpression(indentLevel, n)
 	case parser.TemplElementExpression:
 		err = g.writeTemplElementExpression(indentLevel, n)
+
+		// TemplElementExpression with block should always have whitespace if the next element is also
```
[the remark this answers is not in the document]
Comment (AlexanderArvidsson) [reply generator/generator.go:553]: Great feedback! Let me explain:

The full comment is on that line is:
```
		// TemplElementExpression with block should always have whitespace if the next element is also
		// a TemplElementExpression
```
I wrapped it to keep within a sensible line width.

Templ elements can contain children, which is a block, and is called as such in the code base in tests. I merely used the same terminology.
```
		{
			name: "templelement: simple, block with text",
			input: `@Other(p.Test) {
	some words
}`,
```

When I didn't add this special case, tests were failing. See my comment in one of the outdated threads.
There are tests that expect horizontal spacing between 2 TemplElementExpressions that contain children. Since I did not want to change the test, in order to maintain the original functionality, I added this edge-case, as it was the only way I could figure out to satisfy the existing test.
See test `test-whitespace-around-go-keywords`, which includes multiple Templ elements.
I don't think adding a test for this is necessary since this test already does exactly this, albeit with a dumb name.

The nil check was my bad, missed that one, sorry.

File: generator/generator.go [hunk 4]
```diff
@@ -566,8 +580,10 @@ func (g *generator) writeNode(indentLevel int, current parser.Node, next parser.
 	// Write trailing whitespace, if there is a next node that might need the space.
 	// If the next node is inline or text, we might need it.
 	// If the current node is a block element, we don't need it.
-	needed := (isInlineOrText(current) && isInlineOrText(next))
-	if ws, ok := current.(parser.WhitespaceTrailer); ok && needed {
+	// If, switch and for as current node skip whitespace, but not always when next node.
+	neededWhitespace := forceWhitespace || (maybeWhitespace && isInlineOrText(current) && isInlineOrText(next))
```
[the remark this answers is not in the document]
Comment (AlexanderArvidsson) [reply generator/generator.go:584]: I agree, this whole function is extremely confusion. I did the absolute minimal change to it to allow the existing tests to pass. While I am not happy with it, the truth is that this function needs a huge refactoring. It generalizes adding whitespaces after elements, but I don't agree that its something that can be generalized like this.

I, however, probably don't have the full knowledge of the codebase to write a better one, or argue why a different solution is better.

File: generator/generator.go [hunk 5]
```diff
@@ -528,6 +528,9 @@ func (g *generator) writeNodes(indentLevel int, nodes []parser.Node, next parser
 }
 
 func (g *generator) writeNode(indentLevel int, current parser.Node, next parser.Node) (err error) {
+	maybeWhitespace := true
```
[the remark this answers is not in the document]
Comment (AlexanderArvidsson) [reply generator/generator.go:531]: The intention is the opposite; we if this is true, we might add a whitespace. If it's false, we definitely won't add a whitespace. It's only there to prevent if, for and switch to not add a whitespace, since that's handled elsewhere.

But I agree, your variable names are better.

File: parser/v2/types.go [hunk 6]
```diff
@@ -375,17 +375,27 @@ func (t HTMLTemplate) Write(w io.Writer, indent int) error {
...
 	var hasHorizontalSpace bool
-	for _, r := range s {
+
+	runes := []rune(s)
+
+	for i, r := range s {
 		if r == '\n' {
+			if allowMulti && i < len(runes)-1 {
+				next := runes[i+1]
+				if next == '\n' {
```
[the remark this answers is not in the document]
Comment (AlexanderArvidsson) [reply parser/v2/types.go:395]: I had assumed this was normalized before entering this part of the code. Seems dangerous to have 2 code-paths based on OS everywhere in the code. I'd check, but it'd be difficult for me to test it since I don't have Windows :sweat_smile:

File: generator/generator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexanderArvidsson) [reply generator/generator.go:553]: > if the comment is accurate, with <i>child text node</i><b>next node</b>, we might end up with a line break between those inline nodes, since whitespace would be "forced".

Let me clarify that what it's adding here is a space, not a newline. There are tests that check for this.

**Comment by AlexanderArvidsson** [comment]
Upon adding some more generator tests, I found that the `htmldiff.Diff` incorrectly diffs HTMl:
Consider this expected HTML:
```
<i>child text node</i> <b>next node</b>
```
With this actual HTML:
```
<i>child text node</i><b>next node</b>
```

These two HTMLs are _not the same_. If you open a browser with these two, there is a difference. In HTML, spacing is preserved if 2 inline elements are on the same line. If they're on different lines, an implicit space is added.
However, when `htmldiff.Diff` compares these, it formats them like this:
```
actual:
<i>
 child text node
</i>
<b>
 next node
</b>

expected:
<i>
 child text node
</i>
<b>
 next node
</b>
```

Which are incorrectly marked as equal. I'm not too sure why there's a formatting step here, as the point of generator tests should be to compare the real output with the expected output, no middle-hand formatting.
